File: app/src/main/java/com/soutvoid/gamesproject/interactor/franchise/FranchiseRepository.java
package com.soutvoid.gamesproject.interactor.franchise;

import com.agna.ferro.mvp.component.scope.PerApplication;
import com.soutvoid.gamesproject.domain.franchise.Franchise;
import com.soutvoid.gamesproject.interactor.franchise.network.FranchiseApi;
import com.soutvoid.gamesproject.interactor.util.Fields;
import com.soutvoid.gamesproject.interactor.util.Filter;
import com.soutvoid.gamesproject.interactor.util.Order;
import com.soutvoid.gamesproject.interactor.util.TransformUtil;

import java.util.ArrayList;

import javax.inject.Inject;

import rx.Observable;


@PerApplication
public class FranchiseRepository {

    private FranchiseApi franchiseApi;

    @Inject
    public FranchiseRepository(FranchiseApi franchiseApi) {
        this.franchiseApi = franchiseApi;
    }

    public Observable<ArrayList<Franchise>> searchFranchises(String searchQuery, Fields fields, int limit, int offset, Order order) {
        return franchiseApi.searchForFranchises(fields.toString(), limit, offset, order.toString(), searchQuery)
                .flatMap(franchiseObjs -> Observable.just(TransformUtil.transformCollection(franchiseObjs)));
    }

    public Observable<ArrayList<Franchise>> searchFranchises(String searchQuery,
                                                             Fields fields,
                                                             int limit,
                                                             int offset,
                                                             Order order,
                                                             Filter filter) {
        return franchiseApi.searchForFranchises(fields.toString(), limit, offset, order.toString(), searchQuery, filter.toMap())
                .flatMap(franchiseObjs -> Observable.just(TransformUtil.transformCollection(franchiseObjs)));
    }

    public Observable<ArrayList<Franchise>> getFranchisesById(int id, Fields fields) {
        return franchiseApi.getFranchisesById(id, fields.toString())
                .flatMap(franchiseObjs -> Observable.just(TransformUtil.transformCollection(franchiseObjs)));
    }

}

File: app/src/main/java/com/soutvoid/gamesproject/interactor/franchise/FranchiseRepository.kt
package com.soutvoid.gamesproject.interactor.franchise

import com.agna.ferro.mvp.component.scope.PerApplication
import com.soutvoid.gamesproject.domain.franchise.Franchise
import com.soutvoid.gamesproject.interactor.franchise.network.FranchiseApi
import com.soutvoid.gamesproject.interactor.franchise.network.response.FranchiseObj
import com.soutvoid.gamesproject.interactor.util.*
import rx.Observable
import java.util.*
import javax.inject.Inject


@PerApplication
class FranchiseRepository @Inject
constructor(private val franchiseApi: FranchiseApi) {

    fun searchFranchises(searchQuery: String?, fields: Fields?, limit: Int, offset: Int, order: Order?): Observable<ArrayList<Franchise>> {
        return franchiseApi.searchForFranchises(fields.toString(), limit, offset, order.toString(), searchQuery)
                .flatMap { franchiseObjs -> Observable.just(TransformUtil.transformCollection<Franchise, FranchiseObj>(franchiseObjs)) }
    }

    fun searchFranchises(searchQuery: String?,
                         fields: Fields?,
                         limit: Int,
                         offset: Int,
                         order: Order?,
                         filter: Filter?): Observable<ArrayList<Franchise>> {
        return franchiseApi.searchForFranchises(fields.toString(), limit, offset, order.toString(), searchQuery, filter?.toMap())
                .flatMap { franchiseObjs -> Observable.just(TransformUtil.transformCollection<Franchise, FranchiseObj>(franchiseObjs)) }
    }

    fun getFranchisesById(id: Int, fields: Fields?): Observable<ArrayList<Franchise>> {
        return franchiseApi.getFranchisesById(id, fields.toString())
                .flatMap { franchiseObjs -> Observable.just(TransformUtil.transformCollection<Franchise, FranchiseObj>(franchiseObjs)) }
    }

    fun searchFranchises(query: Query): Observable<ArrayList<Franchise>> {
        return searchFranchises(
                query.searchQuery,
                query.fields,
                query.limit,
                query.offset,
                query.order,
                query.filter)
    }

}
